compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,132 @@
+        //   constructor() { x = ... }
+        //   constructor(...): this() { ... }  // x will be initialized via this() delegation
+
+        // 3) Property can be initialized at any of class's anonymous initializers (all of initializers will be executed), e.g.,
+        //   init { x = ... }
+        //   ...
+        //   init { y = ... }
+
+        // 4) Property can be initialized at other property's initializing expression too, e.g.,
+        //   val initX = inlineMe { x = ... } // where inlineMe returns the value of the last expression of the lambda
+
+        // Here, we accumulate initialized info of a property per constructor, anonymous initializer, or property initialization
+        val initializedInConstructor = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+        val initializedInInit = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+        val initializedInOtherProperty = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+
+        // To handle the delegated constructor call, we need a cache from constructor to (analyzed) property init info.
+        val constructorToData = mutableMapOf<FirConstructor, PropertyInitializationInfo>()
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirProperty, Boolean>,
+            acc: (Boolean, Boolean) -> Boolean,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo =
+                if (delegatedConstructor != null && constructorToData.containsKey(delegatedConstructor)) {
+                    constructorToData[delegatedConstructor]
+                } else null
+
+            val data = PropertyInitializationInfoCollector(memberProperties.map { it.symbol }.toSet()).getData(graph)
+            val infoAtExitNode = data[graph.exitNode]?.get(NormalPath)
+
+            val info = when {
+                delegatedInfo == null && infoAtExitNode == null -> return
+                delegatedInfo == null -> infoAtExitNode
+                infoAtExitNode == null -> delegatedInfo
+                else -> {
+                    // NB: it's not merge, which is conducted at merging points, such as loop condition or when.
+                    // Rather, delegated constructor call is the predecessor of the current constructor call, so we should accumulate.
+                    delegatedInfo.plus(infoAtExitNode)
+                }
+            } ?: return
+
+            if (graph.declaration is FirConstructor) {
+                constructorToData.putIfAbsent(graph.declaration as FirConstructor, info)
             }
+
+            for (property in memberProperties) {
+                if (map.containsKey(property)) {
+                    // Accumulation: && for class constructors, || for class's anonymous initializers and property initializations
+                    map[property] = acc.invoke(map[property]!!, info[property.symbol]?.isDefinitelyVisited() == true)
```

Comment:
Maybe not in this CL, but it seems to be simpler to just store `EventOccurrencesRange` in the map and invoke isDefinitelyVisited at the very end.

Reply:
Oh, this is a good point. It's just a matter of how `op` is defined, and using `EventOccurrenceRange` as values doesn't matter here. I rewrote the maps' values so that it can accumulate event ranges accordingly, and then dragged the `isDefinitelyVisited` computation as late as possible (sounds like `lazy`). Thanks!